iwyu_output.cc

```diff
@@ -1056,8 +1056,11 @@ void ProcessForwardDeclare(OneUse* use,
     }
   }
 
-  // (A7) If --no_fwd_decls has been passed, recategorize as a full use.
-  if (GlobalFlags().no_fwd_decls) {
+  // (A7) If --no_fwd_decls has been passed, recategorize as a full use unless
+  // the decl is in this file (in which case it must be a self-sufficient decl
+  // being used, so we can just let IWYU do its work).
+  if (GlobalFlags().no_fwd_decls &&
+      GetFileEntry(use->decl_loc()) != GetFileEntry(use->use_loc())) {
```

Comment:
This is not the right solution, in my opinion. What if the forward decl is in a different file?
I think we should instead look for the definition of the associated decl, and otherwise leave it as a fwd-decl use. 

Haven't tested it, but something like this should do it:
```cpp
if (GlobalFlags().no_fwd_decls &&
    GetDefinitionForClass(use->decl()) != nullptr) {
...
```

Reply:
We're trying to keep types that *don't* have a definition as forward-declares, but I'm guessing maybe you meant `== nullptr`?

It's an interesting take; you're proposing that if we've said "no fwd decls", we force all uses to full except the ones for which there is no definition available? That sounds good to me, but I think there's a risk that IWYU will suggest to add redundant forward decls to the current file for name-only decls that have leaked into the current translation unit. But it still beats the current behavior. I'll see if I can implement that.

Reply:
I really meant that: The definition for the decl must exist for the fwd-decl use to be convertible to a full use. 
Otherwise, what sense would it make to forcefully convert it to a full use, if the complete type doesn't exist?

Or maybe I am missing something (I havent tried the switch myself)?

Reply:
I applied your suggestion and it passes the current test. I'm not sure what happens with fwd-decls in other files, though. Something like this might go awry:
```
// foohash.h
class Foo;

uintptr_t hash(const Foo* p) {
    return (uintptr_t)p;
}

// foo.h
class Foo {
   int lo;
   int hi;
};

// main.cc
#include "foohash.h"
#include "foo.h"

bool equal(const Foo* a, const Foo* b) {
    return hash(a) == hash(b);
}
```

* main.cc only uses `Foo` in name, and only needs a fwd-decl
* `--no_fwd_decls` stipulates that we don't want fwd-decls, and because `Foo`'s complete type is visible, we promote the uses in `equal` to full uses
* Now we keep foo.h because `Foo` is used, and foohash.h because `hash` is used.

This feels icky, but I think it makes sense. We asked for no fwd-decls. I'll have to test what happens if we remove the `hash` function and let foohash.h carry only the fwd-decl.

Reply:
The same-file check I suggested works better, for some reason. There is other logic earlier in `ProcessForwardDeclare` to ignore the use if the definition is present, I think that could be what helps.

There is logic similar to this in `CalculateIwyuForForwardDeclareUse`, but more sophisticated, but this basic check seems to cover all cases I've found.

I'll push a new version with one more example in the test.

Reply:
@jru Ping! I just rebased this.

I don't remember exactly now why your suggested solution didn't work out, but I couldn't get it to pass the testcase.

Reply:
Sorry, I forgot about this, sorry. I'll take another look!

Reply:
:+1: